runtime/compiler/env/VMJ9.cpp

```diff
@@ -4947,46 +4947,108 @@ TR_J9VMBase::refineInvokeCacheElementSymRefWithKnownObjectIndex(TR::Compilation
    return newRef;
    }
 
+static char *
+getSignatureForLinkToStatic(
+   const char * const extraParamsBefore,
+   const char * const extraParamsAfter,
+   TR::Compilation* comp,
+   J9UTF8* romMethodSignature,
+   int32_t &signatureLength)
+   {
+   const size_t extraParamsLength = strlen(extraParamsBefore) + strlen(extraParamsAfter);
+
+   const char * const origSignature = utf8Data(romMethodSignature);
+   const int origSignatureLength = J9UTF8_LENGTH(romMethodSignature);
+   signatureLength = origSignatureLength + extraParamsLength;
+
+   const int32_t signatureAllocSize = signatureLength + 1; // +1 for NUL terminator
+   char * linkToStaticSignature =
+      (char *)comp->trMemory()->allocateMemory(signatureAllocSize, heapAlloc);
+
+   TR_ASSERT_FATAL(
+      origSignature[0] == '(',
+      "method signature must begin with '(': `%.*s'",
+      origSignatureLength,
+      origSignature);
+
+   // Can't simply strchr() for ')' because that can appear within argument
+   // types, in particular within class names. It's necessary to parse the
+   // signature.
```

Comment:
Is this true, that an argument type or a class name can contain a parenthesis?  That wasn't my understanding, but it has been a while since I've looked at the JLS lexical elements.  Do you have an actual use case of this in hand?  I might learn something new today :-).

Reply:
The only way for an argument type to possibly contain a parenthesis would be via a class name. I don't believe that the JLS lexical syntax determines which characters the JVM accepts in class names. In particular, the binary name of a class almost always contains `/`, which is not allowed in the lexical syntax. (I thought that `$` was another example, used to avoid name collisions, but I was surprised to learn this morning that identifiers in Java source can include it.)

On binary class names, the JVMS (section 4.2.1) says that they
> may be drawn, where not further constrained, from the entire Unicode codespace

and goes on without specifying any further constraints. (I think it's safe to assume that `;` is disallowed though, because `L...;` always relies on the absence of embedded semicolons.) For contrast, the very next section (4.2.2) explicitly forbids `.`, `;`, `[`, `/`, and (most of the time) `<` and `>` in the names of methods, fields, etc.

As I mentioned in the commit message/initial comment, it's hard to tell whether parentheses are *supposed* to be allowed, but I don't see anything forbidding them in the spec. Furthermore, I have been able to get OpenJ9 to load classes whose names contain parentheses, including one that exercises this method with parentheses embedded in the parameter list.

There were some issues I ran into in my experiments, probably due to the lack of clarity around all this. For example, calls to methods with signatures containing embedded parentheses seem to confuse the verifier. During verification, the call arguments are accepted, but the result is assumed to be `Object`, and IIRC it wasn't even possible to then `checkcast` that `Object` to a more specific type. There was also something I tried (I forget what) that caused something in the JVM to claim that the class name was invalid (on I don't know what basis).

Here's [the mentioned class that exercises this case](https://github.com/eclipse-openj9/openj9/files/6423011/StringConcat.Par.en.class.zip). Run:
```
$ java -Xjit:'count=0,limit={*.jitme*}' 'StringConcat(Par)en'
```
Without the fix, I get a segfault. With the fix, it works properly. If I modify `getSignatureForLinkToStatic()` to print both `origSignature` and `linkToStaticSignature`, I can see that it transforms
```
(Ljava/lang/String;LStringConcat(Par)en;I)Ljava/lang/String;
```
into
```
(Ljava/lang/String;LStringConcat(Par)en;ILjava/lang/Object;Ljava/lang/Object;)Ljava/lang/String;
```
which is precisely what I would expect in case the original signature is allowed.

So while it's certainly out of the ordinary, and I wouldn't say that I have a *use case* for such class names, it appears that they can be encountered. Maybe it will turn out later that these names need to be rejected during loading/verification, but the changes in this method will work either way

Reply:
OK.  javac wouldn't let me define a class with a parenthesis, but perhaps one could be manually constructed. 
 Regardless, your code looks good.